bundled/tool/server.py

```diff
@@ -223,6 +252,101 @@ def _get_settings_by_document(document: workspace.Document | None):
 # *****************************************************
 # Internal execution APIs.
 # *****************************************************
+def _explain_check(
+    document: workspace.Document,
+    code: str,
+) -> utils.RunResult | None:
+    """Runs tool on the given document.
+
+    If `use_stdin` is `True` then contents of the document is passed to the tool via
+    stdin.
+    """
+    if str(document.uri).startswith("vscode-notebook-cell"):
+        # Skip notebook cells
+        return None
+
+    if utils.is_stdlib_file(document.path):
+        # Skip standard library python files.
+        return None
+
+    # deep copy here to prevent accidentally updating global settings.
+    settings = copy.deepcopy(_get_settings_by_document(document))
+
+    code_workspace = settings["workspaceFS"]
+    cwd = settings["workspaceFS"]
+
+    use_path = False
+    use_rpc = False
+    if settings["path"]:
+        # 'path' setting takes priority over everything.
+        use_path = True
+        argv = settings["path"]
+    elif settings["interpreter"] and not utils.is_current_interpreter(settings["interpreter"][0]):
+        # If there is a different interpreter set use JSON-RPC to the subprocess
+        # running under that interpreter.
+        argv = [TOOL_MODULE]
+        use_rpc = True
+    elif settings["importStrategy"] == "useBundled":
+        # If we're loading from the bundle, use the absolute path.
+        argv = [
+            os.fspath(pathlib.Path(__file__).parent.parent / "libs" / "bin" / TOOL_MODULE),
+        ]
+    else:
+        # If the interpreter is same as the interpreter running this process then run
+        # as module.
+        argv = [os.path.join(sysconfig.get_path("scripts"), TOOL_MODULE)]
+
+    argv += ["--explain", code, "."]
```

Comment:
We need something that can map a check code to its summary.

Reply:
Implementation I used for the demo

```diff
diff --git a/src/cli.rs b/src/cli.rs
index b1cd519..ed9c7ae 100644
--- a/src/cli.rs
+++ b/src/cli.rs
@@ -109,6 +109,9 @@ pub struct Cli {
     /// The name of the file when passing it through stdin.
     #[arg(long)]
     pub stdin_filename: Option<String>,
+    /// Explain a check.
+    #[arg(long)]
+    pub explain: Option<String>,
 }
 
 impl Cli {
diff --git a/src/main.rs b/src/main.rs
index 5c5e605..d32bfaa 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -1,6 +1,7 @@
 use std::io::{self, Read};
 use std::path::{Path, PathBuf};
 use std::process::ExitCode;
+use std::str::FromStr;
 use std::sync::mpsc::channel;
 use std::time::Instant;
 
@@ -284,6 +285,18 @@ fn inner_main() -> Result<ExitCode> {
         show_settings(configuration, project_root, pyproject);
         return Ok(ExitCode::SUCCESS);
     }
+    if let Some(code) = cli.explain {
+        match CheckCode::from_str(&code) {
+            Ok(code) => {
+                println!("{}", code.kind().summary());
+                return Ok(ExitCode::SUCCESS);
+            }
+            Err(_) => {
+                eprintln!("Unknown check code");
+                return Ok(ExitCode::FAILURE);
+            }
+        }
+    }
 
     // Extract settings for internal use.
     let autofix = configuration.fix;
```

Reply:
Instead, we could provide a CLI option that lists checks ruff supports:

```
ruff --list-errors ...

C401: ...
C402: ...
```

and then parse its output to create a mapping to convert a check code to its summary when the LSP is launched.

Reply:
```diff
diff --git a/src/cli.rs b/src/cli.rs
index b1cd519..764f562 100644
--- a/src/cli.rs
+++ b/src/cli.rs
@@ -109,6 +109,9 @@ pub struct Cli {
     /// The name of the file when passing it through stdin.
     #[arg(long)]
     pub stdin_filename: Option<String>,
+    /// Show all available checks
+    #[arg(long)]
+    pub show_checks: bool,
 }
 
 impl Cli {
diff --git a/src/main.rs b/src/main.rs
index 5c5e605..c0d6d0e 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -26,6 +26,7 @@ use log::{debug, error};
 use notify::{raw_watcher, RecursiveMode, Watcher};
 #[cfg(not(target_family = "wasm"))]
 use rayon::prelude::*;
+use strum::IntoEnumIterator;
 use walkdir::DirEntry;
 
 /// Shim that calls par_iter except for wasm because there's no wasm support in
@@ -284,6 +285,13 @@ fn inner_main() -> Result<ExitCode> {
         show_settings(configuration, project_root, pyproject);
         return Ok(ExitCode::SUCCESS);
     }
+    if cli.show_checks {
+        // Show check summaries.
+        for c in CheckCode::iter() {
+            println!("{}: {}", c.kind().code().as_ref(), c.kind().summary());
+        }
+        return Ok(ExitCode::SUCCESS);
+    }
 
     // Extract settings for internal use.
     let autofix = configuration.fix;
```